File: examples/deno/scripts/agentic_smoke.ts
import { assert, assertEquals } from "@std/assert";

import { bootstrap as bootstrapAgent } from "../apps/agent/app";
import { bootstrap as bootstrapBackend } from "../apps/backend/app";

type RunningServer = Deno.HttpServer<Deno.NetAddr>;

type ApiEnvelope<T> = {
  success: boolean;
  data?: T;
  error?: {
    code?: string;
    message?: string;
  };
};

type Task = {
  id: string;
  title: string;
  description?: string;
  status: "todo" | "doing" | "done";
  createdAt: string;
  updatedAt: string;
};

async function closeServer(server: RunningServer): Promise<void> {
  await server.shutdown();
}

async function startServer(
  kind: "backend" | "agent",
  dataDirectory: string,
): Promise<{
  server: RunningServer;
  baseUrl: string;
}> {
  if (kind === "backend") {
    const app = bootstrapBackend({
      dataDirectory,
      port: 0,
    });
    const server = await app.startBackend();

    return {
      server,
      baseUrl: `http://127.0.0.1:${server.addr.port}`,
    };
  }

  const app = bootstrapAgent({
    dataDirectory,
    port: 0,
  });
  const server = await app.startAgent();

  return {
    server,
    baseUrl: `http://127.0.0.1:${server.addr.port}`,
  };
}

async function parseJson<T>(
  response: Response,
  description: string,
): Promise<T> {
  const payload = (await response.json()) as T;

  if (!response.ok) {
    throw new Error(
      `agentic_smoke: ${description} failed with HTTP ${response.status}`,
    );
  }

  return payload;
}

async function assertSuccess<T>(
  response: Response,
  description: string,
): Promise<T> {
  const payload = await parseJson<ApiEnvelope<T>>(response, description);

  if (!payload.success || payload.data === undefined) {
    throw new Error(
      `agentic_smoke: ${description} failed: ${
        payload.error?.message ?? "missing success payload"
      }`,
    );
  }

  return payload.data;
}

async function assertFailure<T>(
  response: Response,
  description: string,
  expectedStatus: number,
  expectedCode: string,
): Promise<ApiEnvelope<T>> {
  const payload = (await response.json()) as ApiEnvelope<T>;

  assertEquals(
    response.status,
    expectedStatus,
    `agentic_smoke: ${description} must return HTTP ${expectedStatus}`,
  );
  assertEquals(
    payload.success,
    false,
    `agentic_smoke: ${description} must return success=false`,
  );
  assertEquals(
    payload.error?.code,
    expectedCode,
    `agentic_smoke: ${description} must expose ${expectedCode}`,
  );

  return payload;
}

async function run(): Promise<void> {
  const tempDirectory = await Deno.makeTempDir({
    prefix: "app-deno-agentic-smoke-",
  });
  let backendServer: RunningServer | undefined;
  let agentServer: RunningServer | undefined;

  try {
    const backend = await startServer("backend", tempDirectory);
    const agent = await startServer("agent", tempDirectory);
    backendServer = backend.server;
    agentServer = agent.server;

    const catalog = await assertSuccess<{
      systemPrompt: string;
      resources: Array<{ uri: string }>;
      tools: Array<{ publishedName: string; requiresConfirmation: boolean }>;
    }>(
      await fetch(`${agent.baseUrl}/catalog`),
      "agent catalog",
    );

    const toolNames = catalog.tools.map((tool) => tool.publishedName).sort();
    assertEquals(
      toolNames,
      ["task_create", "task_list", "task_move"],
      "agentic_smoke: agent catalog must expose the three task tools",
    );
    assertEquals(
      catalog.systemPrompt.includes("Tool task_create"),
      true,
      "agentic_smoke: /catalog must expose the host global prompt built from the registered tool fragments",
    );
    assertEquals(
      catalog.resources.map((resource) => resource.uri).sort(),
      [
        "app://agent/system/prompt",
        "app://agent/tools/task_create/semantic",
        "app://agent/tools/task_list/semantic",
        "app://agent/tools/task_move/semantic",
      ],
      "agentic_smoke: /catalog must mirror the MCP semantic resources",
    );

    await assertFailure(
      await fetch(`${agent.baseUrl}/tools/task_create/execute`, {
        method: "POST",
        headers: {
          "content-type": "application/json",
        },
        body: JSON.stringify({
          input: {
            title: "   ",
          },
        }),
      }),
      "agent invalid task_create",
      400,
      "VALIDATION_FAILED",
    );

    const created = await assertSuccess<{ task: Task }>(
      await fetch(`${agent.baseUrl}/tools/task_create/execute`, {
        method: "POST",
        headers: {
          "content-type": "application/json",
        },
        body: JSON.stringify({
          title: "Agent-created task",
          description: "Created through apps/agent",
        }),
      }),
      "agent task_create",
    );

    assertEquals(
      created.task.status,
      "todo",
      "agentic_smoke: task_create must create a todo task",
    );

    const listedByAgent = await assertSuccess<{ tasks: Task[] }>(
      await fetch(`${agent.baseUrl}/tools/task_list/execute`, {
        method: "POST",
        headers: {
          "content-type": "application/json",
        },
        body: JSON.stringify({}),
      }),
      "agent task_list",
    );

    assertEquals(
      listedByAgent.tasks.some((task) => task.id === created.task.id),
      true,
      "agentic_smoke: task_list must see tasks created via agent",
    );

    const moveWithoutConfirmation = await fetch(
      `${agent.baseUrl}/tools/task_move/execute`,
      {
        method: "POST",
        headers: {
          "content-type": "application/json",
        },
        body: JSON.stringify({
          taskId: created.task.id,
          targetStatus: "doing",
        }),
      },
    );
    const moveWithoutConfirmationPayload =
      (await moveWithoutConfirmation.json()) as ApiEnvelope<never>;

    assertEquals(
      moveWithoutConfirmation.status,
      409,
      "agentic_smoke: task_move must require confirmation",
    );
    assertEquals(
      moveWithoutConfirmationPayload.error?.code,
      "CONFIRMATION_REQUIRED",
      "agentic_smoke: task_move must expose structured confirmation errors",
    );

    await assertFailure(
      await fetch(`${agent.baseUrl}/tools/task_move/execute`, {
        method: "POST",
        headers: {
          "content-type": "application/json",
        },
        body: JSON.stringify({
          input: {
            taskId: "missing",
            targetStatus: "done",
          },
          confirmed: true,
        }),
      }),
      "agent task_move missing task",
      404,
      "NOT_FOUND",
    );

    const moved = await assertSuccess<{ task: Task }>(
      await fetch(`${agent.baseUrl}/tools/task_move/execute`, {
        method: "POST",
        headers: {
          "content-type": "application/json",
        },
        body: JSON.stringify({
          input: {
            taskId: created.task.id,
            targetStatus: "doing",
          },
          confirmed: true,
        }),
      }),
      "agent task_move",
    );

    assertEquals(
      moved.task.status,
      "doing",
      "agentic_smoke: confirmed task_move must mutate the task",
    );

    const listedByBackend = await assertSuccess<{ tasks: Task[] }>(
      await fetch(`${backend.baseUrl}/tasks`),
      "backend list after agent execution",
    );
    const movedTask = listedByBackend.tasks.find((task) =>
      task.id === created.task.id
    );

    assert(
      movedTask,
      "agentic_smoke: backend must observe tasks mutated by apps/agent",
    );
    assertEquals(
      movedTask.status,
      "doing",
      "agentic_smoke: backend must observe the status written by task_move",
    );

    const backendBurst = 8;
    const agentBurst = 8;
    const concurrentCreates = await Promise.all(
      [
        ...Array.from(
          { length: backendBurst },
          (_, index) =>
            fetch(`${backend.baseUrl}/tasks`, {
              method: "POST",
              headers: {
                "content-type": "application/json",
              },
              body: JSON.stringify({
                title: `Backend burst ${index + 1}`,
              }),
            }),
        ),
        ...Array.from(
          { length: agentBurst },
          (_, index) =>
            fetch(`${agent.baseUrl}/tools/task_create/execute`, {
              method: "POST",
              headers: {
                "content-type": "application/json",
              },
              body: JSON.stringify({
                input: {
                  title: `Agent burst ${index + 1}`,
                },
              }),
            }),
        ),
      ].map(async (request) =>
        await assertSuccess<{ task: Task }>(
          await request,
          "concurrent backend/agent task creation",
        )
      ),
    );

    assertEquals(
      concurrentCreates.length,
      backendBurst + agentBurst,
      "agentic_smoke: concurrent creates must all complete",
    );

    const listedAfterConcurrency = await assertSuccess<{ tasks: Task[] }>(
      await fetch(`${backend.baseUrl}/tasks`),
      "backend list after concurrent backend/agent writes",
    );

    assertEquals(
      listedAfterConcurrency.tasks.length,
      1 + backendBurst + agentBurst,
      "agentic_smoke: concurrent backend/agent writes must preserve every task",
    );

    console.log("agentic_smoke: ok");
  } finally {
    if (agentServer) {
      await closeServer(agentServer).catch(() => undefined);
    }

    if (backendServer) {
      await closeServer(backendServer).catch(() => undefined);
    }

    await Deno.remove(tempDirectory, { recursive: true }).catch(() =>
      undefined
    );
  }
}

void run().catch((error: unknown) => {
  console.error(
    error instanceof Error
      ? error.message
      : "agentic_smoke failed unexpectedly",
  );
  Deno.exit(1);
});

File: examples/flutter/scripts/agentic_smoke.dart
import 'dart:io';

import 'test_support.dart';

Future<void> run() async {
  final tempDirectory = await createTempDataDirectory(
    'app-flutter-agentic-smoke-',
  );
  HttpServer? backendServer;
  HttpServer? agentServer;

  try {
    final backend = await startBackendHost(tempDirectory);
    final agent = await startAgentHost(tempDirectory);
    backendServer = backend.server;
    agentServer = agent.server;

    final catalog = await assertSuccessEnvelope(
      await getJson('${agent.baseUrl}/catalog'),
      'agent catalog',
    );
    final toolNames =
        asJsonMapList(catalog['tools'], 'catalog tools')
            .map((JsonMap tool) => tool['publishedName'].toString())
            .toList(growable: false)
          ..sort();
    expectJsonEqual(toolNames, <String>[
      'task_create',
      'task_list',
      'task_move',
    ], 'agentic_smoke: agent catalog must expose the three task tools');
    expectTrue(
      catalog['systemPrompt'].toString().contains('Tool task_create'),
      'agentic_smoke: /catalog must expose the host global prompt built from the registered tool fragments',
    );
    final resourceUris =
        asJsonMapList(catalog['resources'], 'catalog resources')
            .map((JsonMap resource) => resource['uri'].toString())
            .toList(growable: false)
          ..sort();
    expectJsonEqual(
      resourceUris,
      <String>[
        'app://agent/system/prompt',
        'app://agent/tools/task_create/semantic',
        'app://agent/tools/task_list/semantic',
        'app://agent/tools/task_move/semantic',
      ],
      'agentic_smoke: /catalog must mirror the MCP semantic resources',
    );

    await assertFailureEnvelope(
      await postJson(
        '${agent.baseUrl}/tools/task_create/execute',
        <String, dynamic>{
          'input': <String, dynamic>{'title': '   '},
        },
      ),
      'agent invalid task_create',
      statusCode: 400,
      code: 'VALIDATION_FAILED',
    );

    final created = await assertSuccessEnvelope(
      await postJson(
        '${agent.baseUrl}/tools/task_create/execute',
        <String, dynamic>{
          'title': 'Agent-created task',
          'description': 'Created through apps/agent',
        },
      ),
      'agent task_create',
    );
    final createdTask = asJsonMap(created['task'], 'agent created task');
    expectEqual(
      createdTask['status'],
      'todo',
      'agentic_smoke: task_create must create a todo task',
    );

    final listedByAgent = await assertSuccessEnvelope(
      await postJson(
        '${agent.baseUrl}/tools/task_list/execute',
        const <String, dynamic>{},
      ),
      'agent task_list',
    );
    final listedByAgentTasks = asJsonMapList(
      listedByAgent['tasks'],
      'agent listed tasks',
    );
    expectTrue(
      listedByAgentTasks.any((JsonMap task) => task['id'] == createdTask['id']),
      'agentic_smoke: task_list must see tasks created via agent',
    );

    await assertFailureEnvelope(
      await postJson(
        '${agent.baseUrl}/tools/task_move/execute',
        <String, dynamic>{'taskId': createdTask['id'], 'targetStatus': 'doing'},
      ),
      'agent task_move without confirmation',
      statusCode: 409,
      code: 'CONFIRMATION_REQUIRED',
    );

    await assertFailureEnvelope(
      await postJson(
        '${agent.baseUrl}/tools/task_move/execute',
        <String, dynamic>{
          'input': <String, dynamic>{
            'taskId': 'missing',
            'targetStatus': 'done',
          },
          'confirmed': true,
        },
      ),
      'agent task_move missing task',
      statusCode: 404,
      code: 'NOT_FOUND',
    );

    final moved = await assertSuccessEnvelope(
      await postJson(
        '${agent.baseUrl}/tools/task_move/execute',
        <String, dynamic>{
          'input': <String, dynamic>{
            'taskId': createdTask['id'],
            'targetStatus': 'doing',
          },
          'confirmed': true,
        },
      ),
      'agent task_move',
    );
    final movedTask = asJsonMap(moved['task'], 'agent moved task');
    expectEqual(
      movedTask['status'],
      'doing',
      'agentic_smoke: confirmed task_move must mutate the task',
    );

    final listedByBackend = await assertSuccessEnvelope(
      await getJson('${backend.baseUrl}/tasks'),
      'backend list after agent execution',
    );
    final listedByBackendTasks = asJsonMapList(
      listedByBackend['tasks'],
      'backend tasks after agent execution',
    );
    final backendMovedTask = listedByBackendTasks.firstWhere(
      (JsonMap task) => task['id'] == createdTask['id'],
      orElse: () => fail(
        'agentic_smoke: backend must observe tasks mutated by apps/agent',
      ),
    );
    expectEqual(
      backendMovedTask['status'],
      'doing',
      'agentic_smoke: backend must observe the status written by task_move',
    );

    const backendBurst = 8;
    const agentBurst = 8;
    final concurrentCreates = await Future.wait(<Future<JsonMap>>[
      ...List<Future<JsonMap>>.generate(backendBurst, (int index) async {
        final result = await assertSuccessEnvelope(
          await postJson('${backend.baseUrl}/tasks', <String, dynamic>{
            'title': 'Backend burst ${index + 1}',
          }),
          'concurrent backend create ${index + 1}',
        );
        return asJsonMap(result['task'], 'backend burst task');
      }),
      ...List<Future<JsonMap>>.generate(agentBurst, (int index) async {
        final result = await assertSuccessEnvelope(
          await postJson(
            '${agent.baseUrl}/tools/task_create/execute',
            <String, dynamic>{
              'input': <String, dynamic>{'title': 'Agent burst ${index + 1}'},
            },
          ),
          'concurrent agent create ${index + 1}',
        );
        return asJsonMap(result['task'], 'agent burst task');
      }),
    ]);

    expectEqual(
      concurrentCreates.length,
      backendBurst + agentBurst,
      'agentic_smoke: concurrent creates must all complete',
    );

    final listedAfterConcurrency = await assertSuccessEnvelope(
      await getJson('${backend.baseUrl}/tasks'),
      'backend list after concurrent backend/agent writes',
    );
    expectEqual(
      asJsonMapList(
        listedAfterConcurrency['tasks'],
        'tasks after concurrent backend/agent writes',
      ).length,
      1 + backendBurst + agentBurst,
      'agentic_smoke: concurrent backend/agent writes must preserve every task',
    );

    stdout.writeln('agentic_smoke: ok');
  } finally {
    await closeServer(agentServer);
    await closeServer(backendServer);
    await removeTempDirectory(tempDirectory);
  }
}

Future<void> main() async {
  try {
    await run();
  } catch (error) {
    stderr.writeln(error);
    exitCode = 1;
  }
}
